### PR TITLE
fix(lsp): use the same lsp server for filetype if it exists

### PR DESCRIPTION
**Pull request by sigmaSd**
fix https://github.com/neovim/nvim-lspconfig/issues/2285

### CODE REVIEW

**Comment by sigmaSd** [comment]
I don't know if there is a spec or something to follow, this is just guessing what vscode does.

**Comment by sigmaSd** [comment]
Helix handles this, so I got curious whats the trick, turns out there is no trick, they never spawn an lsp server **in response** to a goToDefinition request, which now I think about actually make sense. Why do we have that code in the first place, are there actual use cases ?

**Comment by sigmaSd** [comment]
If we take a step back, theoretically in response to a gotodefiniton, the lsp server that sent us there should be the one to continue to work

**Comment by sigmaSd** [comment]
It might be only really solvable at neovim level though

**Comment by sigmaSd** [comment]
Nvm, helix does the exact same thing https://github.com/helix-editor/helix/blob/bcdb475b71b0fbabce57344ac8d2575c23b1bbe0/helix-view/src/editor.rs#L944

**Comment by sigmaSd** [comment]
but helix has a sanity check to compare the original lsp that sent us `id` with the one that we found https://github.com/helix-editor/helix/blob/bcdb475b71b0fbabce57344ac8d2575c23b1bbe0/helix-view/src/editor.rs#L958 and it doesn't spawn a different filetype server in any case

This can only be done inside nvim itself

File: lua/lspconfig/configs.lua [hunk 1]
```diff
@@ -245,7 +245,13 @@ function configs.__newindex(t, config_name, config_def)
       end
 
       if root_dir then
-        id = manager.add(root_dir, false)
+        -- Check if there is an active server already for this same file type
+        local active_clients_list = util.get_active_clients_list_by_ft(vim.bo[bufnr].filetype)
+        if #active_clients_list ~= 0 then
+          id = active_clients_list[1].id
```
Comment (glepnir) [inline lua/lspconfig/configs.lua:251]: I don't think there can use index 1 if there has multiple clients for this filetype . like efm tsserver for ts file the 1 can be efm or tsserver .

Comment (sigmaSd) [reply lua/lspconfig/configs.lua:251]: We don't have the original lsp client id at this point, so we can't tell which client to choose

Comment (glepnir) [reply lua/lspconfig/configs.lua:251]: I am trying to remove root_dir in lsp start client in core. maybe can solve this. there is no root_dir definition in lsp doc. use wrokspace folder would be better?

Comment (sigmaSd) [reply lua/lspconfig/configs.lua:251]: I'm not sure if this is a correct approach , helix also have this concept , (emacs as well?)

I'm not sure what vscode does

Comment (sigmaSd) [reply lua/lspconfig/configs.lua:251]: I asked a bit , seems like vscode sets the root workspace as root but it have more logic to add more workspaces (vscode can handle multiple workspaces at the time)

Honestly I can't answer here as I'm not familiar that much with lsp protocol

Comment (glepnir) [reply lua/lspconfig/configs.lua:251]: you mean helix has root_dir ? emacs lsp-bridge, yegappan/lsp  don't have root_dir and the root_dir is not in lsp specification. it just used to check there already has an exist client and reused.

Comment (sigmaSd) [reply lua/lspconfig/configs.lua:251]: That's what I understood when I asked in their matrix channel, seems they're still  designing this though https://github.com/helix-editor/helix/pull/4439

Comment (glepnir) [reply lua/lspconfig/configs.lua:251]: I am not familiar with helix. But this doesn't look like a good way.

Comment (glepnir) [reply lua/lspconfig/configs.lua:251]: maybe we can do it like this ?

```diff
--- a/lua/lspconfig/util.lua
+++ b/lua/lspconfig/util.lua
@@ -253,6 +253,14 @@ function M.server_per_root_dir_manager(make_config)
     -- Check if we have a client already or start and store it.
     if not client_id then
       local new_config = make_config(root_dir)
+      if next(clients) ~= nil then
+        for _, id in pairs(clients) do
+          local client = lsp.get_client_by_id(id)
+          if client.name == new_config.name then
+            return id
+          end
+        end
+      end
       -- do nothing if the client is not enabled
       if new_config.enabled == false then
         return
```

Comment (sigmaSd) [reply lua/lspconfig/configs.lua]: I think that works  I updated the pr with your code

Comment (glepnir) [reply lua/lspconfig/configs.lua]: do we need add the new root dir into the workspaceFolder ?

Comment (sigmaSd) [reply lua/lspconfig/configs.lua]: I'm not sure when I tested with jonhoo original rust problem your snippet fixes it

Comment (glepnir) [reply lua/lspconfig/configs.lua]: maybe them belong to one workspace.

Comment (xiantang) [reply lua/lspconfig/configs.lua]: > maybe them belong to one workspace.

no, they belong different workspace when you try to open another project file

**Pull request merged**